java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/LspSession.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.java.lsp.server;
+
+import java.util.concurrent.Future;
+
+import org.netbeans.api.annotations.common.CheckForNull;
+import org.netbeans.modules.java.lsp.server.debugging.NbProtocolServer;
+import org.netbeans.modules.java.lsp.server.protocol.NbLspServer;
+
+import org.openide.util.Lookup;
+import org.openide.util.lookup.ProxyLookup;
+
+/**
+ * A session that provides associated LSP and DAP servers.
+ *
+ * @author martin
+ */
+public final class LspSession {
+
+    private final ProxyLookup.Controller lspServices = new ProxyLookup.Controller();
+    private final ProxyLookup.Controller dapServices = new ProxyLookup.Controller();
+    private final Lookup sessionLookup;
+
+    private volatile NbLspServer lspServer;
+    private volatile NbProtocolServer dapServer;
+
+    LspSession() {
+        sessionLookup = new ProxyLookup(
+                new ProxyLookup(lspServices),
+                new ProxyLookup(dapServices),
+                Lookup.getDefault()
+        );
+    }
+
+    /**
+     * Set the launched LSP server.
+     */
+    void setLspServer(NbLspServer lspServer) {
+        if (lspServer != null && this.lspServer != null) {
+            throw new IllegalStateException("LSP server is set already");
+        }
+        setServerLookup(lspServer, lspServices);
+        this.lspServer = lspServer;
+    }
+
+    /**
+     * Set the launched DAP server.
+     */
+    void setDapServer(NbProtocolServer dapServer) {
+        if (dapServer != null && this.dapServer != null) {
+            throw new IllegalStateException("DAP server is set already");
+        }
+        setServerLookup(dapServer, dapServices);
+        this.dapServer = dapServer;
+    }
+
+    private static void setServerLookup(ScheduledServer server, ProxyLookup.Controller lookupControler) {
```

Comment:
Q: if this is called first for DAP, then for LSP - will the `server.getServerLookup()` contain both DAP + LSP services ?

Reply:
The order does not matter. But lookups of individual servers need to contain just their lookups. The joined lookup is returned from `LspSession.getLookup()`.

Reply:
Sorry, I was not reading well the diff (I thought there's just one controller).